flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/benchmark/StreamNetworkBenchmarkEnvironment.java

```diff
@@ -270,4 +266,18 @@ private InputGateDeploymentDescriptor createInputGateDeploymentDescriptor(
 			consumedSubpartitionIndex,
 			channelDescriptors);
 	}
+
+	private InputGate createInputGateWithMetrics(
+		SingleInputGateFactory gateFactory,
+		InputGateDeploymentDescriptor gateDescriptor,
+		int channelIndex) {
+
+		final SingleInputGate singleGate = gateFactory.create(
+			"receiving task[" + channelIndex + "]",
+			gateDescriptor,
+			SingleInputGateBuilder.NO_OP_PRODUCER_CHECKER,
+			InputChannelTestUtils.newUnregisteredInputChannelMetrics());
+
+		return new InputGateWithMetrics(singleGate, new SimpleCounter());
```

Comment:
Could we somehow deduplicate the code of wrapping the input gate between production and benchmarking code? So that they follow the same path? For example moving the wrapping for both code paths to the `SingleInputGateFactory`?

Reply:
Yes, we could make `SingleInputGateFactory` always return wrapped `InputGateWithMetrics`

Reply:
There has one concern here. 
`SingleInputGateFactory` is `NetworkEnvironment`/`ShuffleService` specific, not a general factory. `InputGateWithMetrics` is not belonging to `ShuffleService`, and it should be in task scope. So it might be not proper to let `SingleInputGateFactory` wrap for task related components, otherwise it seems no need to introduce `InputGateWithMetrics` in this PR. WDYT?

Reply:
I agree that `InputGateWithMetrics` does not belong Netty implementation of shuffle service this is why we are doing it change to keep this notification concern outside.

Reply:
Ok, I understand. In that case maybe we could consider some another layer of factory/builder? One that takes a concrete NetworkEnvironment and creates InputGates wrapped with metrics? I'm not sure if that's worth, but you see the code duplication here? And the risks that if those two code parts will drift apart, the benchmarks might become inaccurate.

Reply:
I could understand your concern here. 
ATM the `InputGateWithMetrics` is only created in the benchmark which would reflect the real process. Considering this wrapper is very easy to construct and the duplication here is not obvious, just a local `singleGate` definition, I think we could introduce another factory for this wrapped `InputGate` if necessary later?

Reply:
Ok